JSON backend

franzpoeschel opened the pull request:
Implementation of a JSON backend for openPMD.

Todo/Issues:
* [x] user documentation
* [x] Some of the methods in `AbstractIOHandlerImpl.hpp` did not behave exactly as documented (modulo misunderstandings on my side). I will address this separately.
* [ ] see the discussion in #65 
* [x] integration with the build system: #384
* [x] remove commits from #337 and rebase onto master
* [x] put appropriate license, copyright and author headers in every file

franzpoeschel committed:
JSON backend

Implement JSON backend to dummy state

Misses the actual implementation of AbstractIOHandlerImpl

Declare IOHandlerImpl for JSON and integrate it with other places

Misses the implementation.

Undebugged minimum implementation for JSON writing

First basically runnable version of JSON writing

To address:
No reading or deleting yet.
Datatypes are currently ignored and the data is assumed to be int64_t.
Attribute values are ignored and replaced with a dummy value.
If a subgroup name can be parsed as a nonnegative string, the JSON API
will create a JSON array rather than a JSON object (associative array) as intended.

Correctly handle groups that can be parsed as int

See last commit's description.

Fix index calculation with offsets in WriteData

Fix some mistakes in JSON writing

Correctly handle overwriting files:
-> overwritten files should not be possible to access any longer and be
clearly distinguished from the newly-created file
Make some verifications execute independent of compiler options.

Full implementation of JSON writing

Respects all datatypes now.

Format code according to Clion Stylesheet

https://github.com/ComputationalRadiationPhysics/contributing/blob/master/IDESettings/CLion/CRP_CLion2016_1.xml

Add generic branching over an openPMD datatype

First runnable version of JSON Reading

Cleanup and implementation of dataset extension

Undebugged version of JSON deletion

Properly (de)serialize datatypes

Instead of casting the Datatype enum to and from int (which is likely to break
when altering the enum), serialize to and from String values.

Fix a number of mistakes in JSON reading and writing

Cleanup

Add JSON tests and fix bugs found thusly

Add further tests and fix a further bug

The JSON library does not understand long double values (i.e. 128bit
floats), represent them as a char array.

Handle floating point special values

JSON represents +/-Infinity and NaN values as null. The JSON library
will correctly serialize those values *to* JSON, implement
(semi)-correct handly for deserialization. As it is unclear which exact
value a null represents, deserialize it to NaN.
Take notice that large floating point values (128 bit) might be
serialized to null as well.

Use std::is_floating_point to distinguish them from other types

Additionally write the byte width of the underlying type

Not yet used in reading

Mark the writable written after successfully extending a dataset

Remove support for absolute paths from openPath

Fix some rough edges from rebasing

Add documentation for the JSON backend

Integrate the JSON backend with the build system

Make platform bytewidth information global per JSON file

Was previously annotated for every single Dataset and Attribute.

Fixes

-> shadowed variables
-> return file streams by shared pointer so that Clang won't be mad
about missing copy constructors
-> handle inconsistencies between GCC, Clang and MSVC in switchType
function

Add licensing information and cleanup imports

Cleanup

Remove unused code from Filesystem.(h|c)pp
Unify replace_all and replace_all_nonrecursively in StringManip.hpp

Move implementation of JSONFilePosition class to source file

Formatting

Refactor tests

Move tests into functions of the backend to be used.

Remove /subgroups/ and /datasets/ indirection

Make optional_paths_110_test HDF5 specific again

Eliminate "extent" field from datasets

Remove object_type key

ax3l commented:
@franzpoeschel I know the changes were quite heavy recently (big sorry again). Can you please try the following to update this branch? I would like to remove the number of commits and would like to have no merge commit in it.

```bash
# backup in case things go south
git checkout -b json-backend-bak
# switch back to json-backend 
git checkout -

# rebase: squash
git fetch --all
# I assume your "git remote -v" for this mainline repo is called "mainline"
git rebase -i mainline
# an editor will open
#   change "pick" to "s" ("squash") to all your commits *but the first entry*
#   remove the lines with "merge commits",
#   remove commits a7a7974 and 176d37d 
# save
# solve merge conflicts one

# if anything goes very wrong
# git rebase --abort

# in the end, update the PR via
git push -f
```

franzpoeschel commented:
Concerning user documentation, is there a place where the file formats are / should be documented?

@ax3l Done, thanks for the steps.

ax3l commented:
>Concerning user documentation, is there a place where the file formats
>are / should be documented?

Yes, it's even recommended :) You can update the documentation files (.rst) in the same PR that lands the feature :)

ax3l commented:
@franzpoeschel I added complete build system support for you in #384.
Is it ok if I rebase this branch and force-push update it against those changes?

ax3l commented:
Pushed the rebased branch :)
Please make sure that you don't have uncommited/unpushed *local* changes before continuing to work on this. Fetch the latest online version, revert the last local commit if they were identical before my rebase and pull the fresh commit.

```bash
git status -sb
# nothing uncommited?
# latest commits were pushed online?

# assuming yes

# fetch the updates I pushed on your branch and remove local "outdated" commit
git fetch --all
git reset --hard HEAD~
git pull --ff-only
```

After that, you can continue as usual.

If you *do* have local changes that were not yet pushed - commit them and rebase the branch against the online commit (also removing the one outdated commit but keeping all other changes on top).

```bash
git fetch --all

git rebase -i mainline/dev
# remove outdated commit
# keep all others
```

franzpoeschel commented:
Thanks for the help and the detailed instructions!

ax3l commented on include/openPMD/auxiliary/Memory.hpp:
```diff
@@ -24,8 +24,10 @@
 #include "openPMD/Datatype.hpp"
 
 #include <functional>
+#include<iostream>
```
little formatting issue :)

franzpoeschel replied:
Ah, good you spotted that, those includes were left from when I added a function to that file which I later removed again. Removed the includes.

ax3l commented on include/openPMD/auxiliary/Filesystem.hpp:
```diff
@@ -87,5 +87,27 @@ remove_directory(std::string const& path);
 bool
 remove_file(std::string const& path);
+
+// /** Opens the file identified by the given path.
```
can this  dead code be removed? :)

franzpoeschel replied:
done ;)

ax3l commented on include/openPMD/IO/JSON/JSONFilePosition.hpp:
```diff
@@ -0,0 +1,53 @@
+    struct JSONFilePosition :
+        public AbstractFilePosition
+#if openPMD_HAVE_JSON
+    {
```
can you please implement the bidy in a `.cpp` file? Avoids pulling in the public `nlohmann/json.hpp` header into our public API.

franzpoeschel replied:
Done.... but I still need to include `nlohmann/json.hpp` in the header file?
This being said, `JSONFilePosition.hpp` is only ever included in `JSONIOHandlerImpl.hpp`, which is included in `JSONIOHandler.hpp` and that header file is only included in `*.cpp` files, so it should be fine?

ax3l replied:
No, if no symbols/declarations are used from `nlohmann/json.hpp` in the header file, then do not include it there.

> This being said, [...]

That sounds good, I wasn't sure.

franzpoeschel replied:
>if no symbols/declarations are used from nlohmann/json.hpp in the header file

They are. `JSONFilePosition.hpp` is implemented as a wrapper around `json::json_pointer`.

ax3l commented on include/openPMD/Datatype.hpp:
```diff
@@ -518,6 +519,184 @@ isSame( openPMD::Datatype const d, openPMD::Datatype const e )
+ * the passed arguments and the template parameter type corresponding to the
+ * openPMD type.
+ */
+
```
maybe no empy newline here. You can put the `OPENPMD_TEMPLATE_OPERATOR` define block above the doxygen block

franzpoeschel replied:
done

ax3l commented on docs/source/backends/json_example.json:
```diff
@@ -0,0 +1,154 @@
+{
```
When we are finished with the review (until then, pls update here):
Shall we rather put this file into https://github.com/openPMD/openPMD-example-datasets ?

ax3l commented on docs/source/backends/json.rst:
```diff
@@ -0,0 +1,82 @@
+The example code in the :ref:`usage section <usage-serial>` will produce the following JSON serialization
+when picking the JSON backend:
+
+.. literalinclude:: json_example.json
```
maybe link to https://github.com/openPMD/openPMD-example-datasets when it's included there?

ax3l commented on docs/source/backends/json_example.json:
```diff
@@ -0,0 +1,154 @@
+  },
+  "subgroups": {
+    "data": {
+      "subgroups": {
```
why do you add a `subgroups` directory here and before `data`? Can we skip those maybe?

franzpoeschel replied:
Short answer: separating namespaces. If this indirection were dropped, the subgroups' and datasets' names will have the same namespace as all other JSON objects at that level, including the objects `attributes` and for the top-level group `platform_byte_widths` (meaning you can't have groups with those names). This seems like a nasty corner case for a user of the API to run into, hence I chose clarity at the expense of a certain informational overhead as is explained in [here](https://github.com/openPMD/openPMD-api/blob/ab2fb040060089a72acaa2c14a8c72a909df2b66/docs/source/backends/json.rst).
Do you have a better solution? I can imagine wrapping the subgroups into quotes or maybe reserving a prefix for "keywords" (such as `__attributes` maybe). Would this be preferrable?

ax3l replied:
Yes, there is a risk of collisions which in practice should not be too often the case. Nevertheless, rather adding a *dict of attributes* inside an `attributes` group would probably be best (no `__`). This keeps the collisions minimal and at the same time is very readable. But that's already the case, no?

The same is true for data sets and groups. Without the `subgroups` "namespace" one can not write the same group and data set name in the same level and neither can be named "attributes". But imho that's fine.

franzpoeschel replied:
>But that's already the case, no?

yep

I changed it now, so attributes keep their separate namespace but datasets and subgroups are now directly linked into the containing group's object, see [here](https://github.com/franzpoeschel/openPMD-api/blob/json-backend/docs/source/backends/json_example.json).

ax3l replied:
Looks good!
Do you think you need the `"object_type": "dataset"` and `"object_type": "group"` keys? They don't hurt, but I wonder if one can already guess by the name (e.g. `data`) and content already.

franzpoeschel replied:
I added those for the `listPaths` and `listDatasets` operations. I did not want to rely on checking the structure of JSON objects to distinguish groups from datasets, so I made the distinction explicit.
I mean, both layouts (i.e. of groups vs. datasets) should actually be computationally distinguishable, but should we go that length?

ax3l replied:
I get the first part, but think it's probably ok to check if a `data` members is present.

Can please you elaborate what you mean with the last sentence (`computationally distinguishable`)?

franzpoeschel replied:
>Can please you elaborate what you mean with the last sentence (computationally distinguishable)?

Pretty much what you said already, datasets and groups can be distinguished by inspecting their members.
>think it's probably ok to check if a data members is present

Probably not a good idea if you look at [line 28](https://github.com/franzpoeschel/openPMD-api/blob/json-backend/docs/source/backends/json_example.json) of the example JSON file ;)
The safest bet would probably be to check whether the `extent` key exists *and* whether it points to a JSON array. That should work, but it *feels* like a hack. Should I do it this way?

ax3l replied:
Why do we need `extent`? Can't we determine this from a list (of lists (of lists)) by querying their lengths, even in ND-cases?

Would checking for "contains a data member that contains a list" be enough to detect a low-level "data set" vs. low-level "group"?

I just want to avoid introducing too many non-standard attributes/sub-groups/naming conventions. It might be that we need some, since we don't have as in XML/HDF5/ADIOS/... attributes on a path that are distinct from attributes on data sets. But we should keep it minimal, e.g. for attributes it's reasonable to have this sub-group and also to have a "data" key.

franzpoeschel replied:
>Why do we need extent? Can't we determine this from a list (of lists (of lists)) by querying their lengths, even in ND-cases?

In principle, yes. It would just require pre-filling the dataset with `null` values to be sure that it does actually have the specified dimensions. (Currently, the data array is just as big as necessary for holding the chunks that have actually been written).
Would you prefer this?

franzpoeschel replied:
The latest commit eliminates the `extent` field, next up will be the elimination of `object_type`.

ax3l replied:
> pre-filling the dataset with null values

makes probably sense, this only works in serial anyway, right?

ax3l replied:
> The latest commit eliminates the extent field, next up will be the elimination of object_type.

Can you please update the checked-in example file? :)

franzpoeschel replied:
>this only works in serial anyway, right?

right
`object_type` is no more and the example file and documentation are updated.

ax3l replied:
cool, thanks! looks good :)

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -2296,3 +2296,475 @@ TEST_CASE( "no_serial_adios1", "[serial][adios]")
+
+#if openPMD_HAVE_JSON
+
+TEST_CASE( "json_write_test", "[serial][json]")
```
Do you think we can just generalize the `hdf5_`/`adios_`/`json_` write test into a helper function and loop with a `Series` object for each backend? Otherwise we copy so much code...

franzpoeschel replied:
Pushed a commit that does this. This leaves only the tests using samples as backend-specific tests.

ax3l reviewed (commented):
Dang, I reviewed this a week ago and forgot to submit my comments.

ax3l commented:
@franzpoeschel I think I can implement #389 now :)
That means: when it is merged you can remove the changes you made to the vendored `nlohmann-json` files, which is a good thing. If you like, rebase and squash some commits when you are on it? (But again, wait for both until #389 is merged) :)

ax3l commented:
@franzpoeschel ready for rebase, remove of the JSON third party patch and some squashing of commits :)

ax3l commented:
@franzpoeschel just do (in case your `git remote` for the main repo is called `mainline`, see `git remote -v`):
```bash
git fetch mainline
git rebase -i mainline/dev
git push -f
```

ax3l commented:
Ouch, looks like #389 did not trigger `-isystem` yet... (Sorry!)

ax3l commented:
@franzpoeschel I don't know why this did not work, but you can fix it by adding an extra commit changing the following:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2c73e2d..418428f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -127,8 +127,10 @@ else()
 endif()
 if(openPMD_HAVE_JSON)
     add_library(openPMD::thirdparty::nlohmann_json INTERFACE IMPORTED)
-    target_link_libraries(openPMD::thirdparty::nlohmann_json
-        INTERFACE nlohmann_json::nlohmann_json)
+    #target_link_libraries(openPMD::thirdparty::nlohmann_json
+    #    INTERFACE nlohmann_json::nlohmann_json)
+    get_target_property(lib_include_dirs nlohmann_json::nlohmann_json INTERFACE_INCLUDE_DIRECTORIES)
+    target_include_directories(openPMD::thirdparty::nlohmann_json SYSTEM INTERFACE ${lib_include_dirs})
 endif()
 
 # external library: HDF5 (optional)
```

ax3l committed:
Nlohmann-JSON: isystem

Should have been automatically since wrapped in an `IMPORTED`
target... but for some reason did include still with `-I`.

ax3l commented:
I pushed it for you.

ax3l reviewed (approved):
Hurray, thanks a lot this is great! :sparkles:
